src/Model/Operation/ProductSyncHandler.php

```diff
@@ -145,18 +148,25 @@ private function doUpsertOperation(
             $channelId,
         );
 
-        /** @var SalesChannelProductEntity $product */
+        /** @var ProductEntity $product */
         foreach ($productCollection as $product) {
             // TODO: up to 2MB payload!
             $nostoProducts = [];
             foreach ($this->processProductVariants($product, $context) as $handledProduct) {
-                $nostoProducts[] = $this->handleProduct(
-                    $handledProduct,
-                    $context,
-                    $account,
-                    $hideProductsAfterClearance,
-                    $ids,
-                );
+                $shopwareProduct = $this->getShopwareProduct($handledProduct->id, $context);
```

Comment:
Wouldn't it make more sense to switch to the `EntityRepository` in general? This way, we would increase the export time, because we fetch each product another time.

Reply:
I'm not sure if the `ProductEntity` contains all the necessary properties. As I mentioned earlier, it lacks the properties `calculatedPrices` and `calculatedPrice`, which are used in the Builder class. You pointed out that we use a calculator similar to the Findologic plugin. I tried that approach, but then I found that we also lack the `cheapestPrice` property, which is necessary for the `calculateCheapestPrice` method in the Shopware class `ProductPriceCalculator`.

Reply:
The calculator also calculates the cheapest price:
https://github.com/shopware/shopware/blob/33829c91107f8acd48fb791a66d360db9532d194/src/Core/Content/Product/SalesChannel/Price/ProductPriceCalculator.php#L53

Reply:
Yes, but before that, the cheapestPrice property must exist. This property doesn't exist in the ProductEntity.
https://github.com/shopware/shopware/blob/33829c91107f8acd48fb791a66d360db9532d194/src/Core/Content/Product/SalesChannel/Price/ProductPriceCalculator.php#L116

Reply:
![image-20240313-103039](https://github.com/Nosto/nosto-shopware6/assets/153912622/d74b1dc9-12e3-4f8f-b2b1-1fd5a68f6991)

Reply:
Yeah, I just saw that we also fetch the sales channel entity within the Findologic plugin, but it's just not typehinted like that.
Im this case, we can go with the sales channel repository. But please fetch it for all the handled products at once. Otherwise there is a database query for each of them.

Reply:
More or less, yes:
1. Save products from `processProductVariants` into a variable.
2. Request all products as Shopware products in one query. (Function should be renamed to `getShopwareProducts`)
3. Access the collection within the `foreach`

It probably also makes sense to change the return type and logic for `processProductVariants` to a `ProductCollection`, so it simplifies the further handling.

Reply:
Is it possible to retain the children of the product that are set in the `processProductVariants` method when we pass those products to the `getShopwareProducts` method?

Reply:
I don't think so

Reply:
I suppose then that it's best to delete the products that are not needed at the end of the `processProductVariants` method?

Reply:
Not sure if I completely get what you mean, but just saw that calling `deleteVariantProducts` and `doDeleteOperation` outside of the foreach in one call would be great as well.

The preservation of the children will probably need to stay the same, but I don't see an issue with that.

Reply:
Example: We have a product that is inactive and has no children. It will be returned as it is by the `processProductVariants` method. Furthermore, if we pass it along with other active products to the `getShopwareProducts` method, it will not be included in the array to be returned because it does not meet the criteria. In this case, it should be deleted. This seems like a complicated solution because we would have to compare two arrays to delete those that are missing in the array returned by the `getShopwareProducts` method.

Also, when we send the product IDs to the `getShopwareProducts` method, all of them will have children set to null. That's why I had to use `$shopwareProduct->setChildren($handledProduct->getChildren());` to set the children that were previously set in the `processProductVariants` method.

Reply:
> We have a product that is inactive and has no children. It will be returned as it is by the processProductVariants method.

As already mentioned in another comment, I don't think this should be the case and it was already changed within one of my suggestions. `processProductVariants` should not include inactive products with no variants. If `handleFirstActiveVariant` returns `null`, you can already delete the product within `processProductVariants`.

Why would something like this not work:
```php
// change return type to ProductCollection
$handledProducts = $this->processProductVariants($product, $context);
$shopwareProducts = $this->getShopwareProducts($handledProducts->getIds(), $context);

foreach ($handledProducts as $handledProduct) {
    $shopwareProduct = $shopwareProducts->get($handledProduct->getId());

    if ($shopwareProduct) {
        $shopwareProduct->setChildren($handledProduct->getChildren());
        $nostoProducts[] = $this->handleProduct(
            $shopwareProduct,
            $context,
            $account,
            $hideProductsAfterClearance,
            $ids,
        );
    } else {
        $this->deleteVariantProducts($handledProduct, $context, $account, $ids);
        $this->doDeleteOperation(
            $account,
            $context,
            [$handledProduct->getId(), $handledProduct->getParentId()],
            $ids
        );
    }
}
```